src/main/java/org/mule/extension/db/internal/source/RowListener.java

```diff
@@ -0,0 +1,219 @@
+import org.mule.runtime.extension.api.annotation.param.NullSafe;
+import org.mule.runtime.extension.api.annotation.param.Optional;
+import org.mule.runtime.extension.api.annotation.param.Parameter;
+import org.mule.runtime.extension.api.annotation.param.ParameterGroup;
+import org.mule.runtime.extension.api.annotation.param.display.DisplayName;
+import org.mule.runtime.extension.api.annotation.param.display.Summary;
+import org.mule.runtime.extension.api.runtime.operation.Result;
+import org.mule.runtime.extension.api.runtime.source.PollContext;
+import org.mule.runtime.extension.api.runtime.source.PollContext.PollItem;
+import org.mule.runtime.extension.api.runtime.source.PollingSource;
+import org.mule.runtime.extension.api.runtime.source.SourceCallbackContext;
+
+import java.io.Serializable;
+import java.util.List;
+import java.util.Map;
+import java.util.function.BiConsumer;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Selects from a table at a regular interval and generates one message per each obtained row.
+ * <p>
+ * Optionally, watermark and id columns can be provided. If a watermark column is provided, the values taken from that column
+ * will be used to filter the contents of the next poll, so that only rows with a greater watermark value are returned. If an
+ * id column is provided, this component will automatically make sure that the same row is not picked twice by concurrent polls
+ *
+ * @since 1.3
+ */
+@MetadataScope(outputResolver = RowListenerMetadataResolver.class)
+@DisplayName("On Table Row")
+@Summary("Triggers a message per each row in a table")
+@Alias("listener")
+public class RowListener extends PollingSource<Map<String, Object>, Void> {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(RowListener.class);
+
+  /**
+   * The name of the table to select from
+   */
+  @Parameter
+  @MetadataKeyId
+  private String table;
+
+  /**
+   * The name of the column to use for watermark
+   */
+  @Parameter
+  @Optional
+  private String watermarkColumn;
+
+  /**
+   * The name of the column to consider as row ID. If provided, this component will make sure that the same row is not
+   * processed twice by concurrent polls.
+   */
+  @Parameter
+  @Optional
+  private String idColumn;
+
+  @ParameterGroup(name = ADVANCED)
+  @NullSafe
+  private QuerySettings settings;
+
+  @Config
+  private DbConnector config;
+
+  @Connection
+  private ConnectionProvider<DbConnection> connectionProvider;
+
+  private final QueryResolver<ParameterizedStatementDefinition> queryResolver = new ParameterizedQueryResolver<>();
+  private ItemHandler idHandler;
+  private ItemHandler watermarkHandler;
+
+  @Override
+  protected void doStart() throws MuleException {
+    if (idColumn != null) {
+      idHandler = (item, row) -> {
+        Object id = row.get(idColumn);
+        if (id != null) {
+          item.setId(id.toString());
+        }
```

Comment:
`else if (LOGGER.isDebugEnabled()) { LOGGER.debug(format(A null watermark value was obtained for row %s ....))}`